### ISSUE TITLE
Merging duplicate markers

### ISSUE THREAD

**Issue by drewnoakes** [svg/svgo#802]
An Inkscape document I created used mid-markers on paths.

After running SVGO I see:

```xml
<defs>
  <marker id="b" orient="auto" overflow="visible">
    <path fill="#727272" fill-rule="evenodd" stroke-width="1pt" d="M-2.5-1c0 2.76-2.24 5-5 5s-5-2.24-5-5 2.24-5 5-5 5 2.24 5 5z" transform="matrix(.2 0 0 .2 1.48 .2)"/>
  </marker>
  <marker id="b" orient="auto" overflow="visible">
    <path fill="#727272" fill-rule="evenodd" stroke-width="1pt" d="M-2.5-1c0 2.76-2.24 5-5 5s-5-2.24-5-5 2.24-5 5-5 5 2.24 5 5z" transform="matrix(.2 0 0 .2 1.48 .2)"/>
  </marker>
  ...
</defs>

<path marker-mid="url(#b)" ... > ... </path>
<path marker-mid="url(#c)" ... > ... </path>
...
```

This particular document has 15 duplicate marker definitions. Merging them reduces the file size by 30%.

Would merging these duplicate markers be a safe optimisation for SVGO? Are there similar examples I can look at to help build a PR?

As I said elsewhere, I'm interested in making some contributions to this project. I'd like some feedback on ideas before putting together PRs.

**Comment by GreLI** [comment]
I guess, it can be. However, I think it's a rare case, so it will not give much profit in general. There is no same plugin. [mergePaths](https://github.com/svg/svgo/blob/a7f62d89846cc94542b75a7a9e1fbd2a4dd64fa4/plugins/mergePaths.js) concatenates sibling paths with same params, but you'll have to track IDs.

**Comment by drewnoakes** [comment]
@GreLI thanks for your responses.

I suppose there are two general scenarios that could be looked at here:

1. Merge duplicate defs
2. Remove unused defs

Both will require the ability to track IDs, as you say.

My questions are:

- Is the removal of unused defs already implemented? (I didn't find it when I looked)
- Is there any guidance on how defs can be referenced/discovered elsewhere in the document? (e.g. specific attributes, specific syntax, whole/partial attribute value matching on `url\(#[^)\s]+\)` or similar)

**Comment by drewnoakes** [comment]
[`cleanupIDs.js`](https://github.com/svg/svgo/blob/master/plugins/cleanupIDs.js#L64) discovers IDs and can serve as a reference for this.

**Comment by GreLI** [comment]
Yes, useless defs are being removed by `removeUselessDefs`. `cleanupIDs` removes unused IDs, so it removes defs without IDs.
No guidance for that for now. Only `cleanupIDs` saves IDs for its internal need.s

**Comment by iology** [comment]
I want this features. These duplicates are usually generated by copy&paste.

**Comment by flipflop97** [comment]
Yes please, Inkscape creates lots of duplicate defs.

**Comment by guignonv** [comment]
+1. I do a lot of copy/paste and generate a huge lot of duplicate defs. My document takes time to load and the list of markers becomes full of the same markers! :-S

**Comment by PRR24** [comment]
My +1, very useful to remove duplicate raster images (for exmaple created by inkscape or poppler), significantly reducting some svg-s.

**Comment by nschloe** [comment]
I also find SVGs in the wild with _excessive_ use of duplicate `<def>`s. Most of the time those are simple red lines like

```xml
<defs>
  <image ns1:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAABCAIAAABR8BlyAAAABmJLR0QA/wD/AP+gvaeTAAAADklEQVR42mO4lZZGNgIA9wYg+cZ4RWoAAAAASUVORK5CYII=" id="f" width="20" height="1" x="0" y="0" />
...
</defs>
```

with hundreds of repetitions.

![b](https://github.com/svg/svgo/assets/181628/75afd9d0-d387-4aba-9bf4-71ffd960707f)
